lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1712,6 +1756,94 @@ class SCFToCalyxPass : public SCFToCalyxBase<SCFToCalyxPass> {
 private:
   LogicalResult partialPatternRes;
   std::shared_ptr<calyx::CalyxLoweringState> loweringState = nullptr;
+
+  FuncOp createNewTopLevelFcn(ModuleOp moduleOp, std::string &baseName) {
+    std::string newName = baseName;
+    unsigned counter = 0;
+    while (SymbolTable::lookupSymbolIn(moduleOp, newName)) {
+      newName = baseName + "_" + std::to_string(++counter);
+    }
+
+    OpBuilder builder(moduleOp.getContext());
+    builder.setInsertionPointToStart(moduleOp.getBody());
+
+    FunctionType funcType = builder.getFunctionType({}, {});
+
+    if (auto newFunc =
+            builder.create<FuncOp>(moduleOp.getLoc(), newName, funcType)) {
+      baseName = newName;
+      return newFunc;
+    }
+
+    moduleOp.emitError("Cannot create new top-level function.");
```

Comment:
```suggestion
    return moduleOp.emitError("Cannot create new top-level function.");
```

Reply:
Not really, the function signature is `FuncOp createNewTopLevelFn...`. We can't really return `emitError`

Reply:
You're right, apologies; I was thinking about LogicalResult or InFlightDiagnostic, where [you can do something like this](https://github.com/llvm/llvm-project/blob/39f78464d5f8794bc836b5e15c92fd1e162a3d0b/mlir/include/mlir/Dialect/Transform/Interfaces/TransformInterfaces.h#L1601).